File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.cpp
class Solution {
public:
    int Possible(vector<int>b,int d,int k){
        int count=0;
        int noOfBloom=0;
        for(int i=0;i<b.size();i++){
            if (b[i]<=d){
                count++;
            }else{
                noOfBloom+=(count/k);
                count=0;
            }
            
        }
        noOfBloom+=(count/k);
        return noOfBloom;

    }
    int minDays(vector<int>& bloomDay, int m, int k) {
        if ((long long)m*k>bloomDay.size()){
            return -1;
        }
        int low=*min_element(bloomDay.begin(), bloomDay.end());
        int high=*max_element(bloomDay.begin(), bloomDay.end());
        int ans=high;
        while(low<=high){
            int mid=(low+high)>>1;
            if (Possible(bloomDay,mid,k)>=m){
                ans=mid;
                high=mid-1;
            }else{
                low=mid+1;
            }
        }
        return ans;
    }
};

File: 1482-minimum-number-of-days-to-make-m-bouquets/1482-minimum-number-of-days-to-make-m-bouquets.py
class Solution(object):
    def Valid(self,b,d,k):
        count,NoOfBloom=0,0
        for i in b:
            if(i<=d):
                count+=1
            else:
                NoOfBloom+=(count//k)
                count=0
        NoOfBloom+=(count//k)
        return NoOfBloom
    
    def minDays(self, bloomDay, m, k):
        """
        :type bloomDay: List[int]
        :type m: int
        :type k: int
        :rtype: int
        """
        if(m*k>len(bloomDay)):
            return -1
        low,high=min(bloomDay),max(bloomDay)+1
        ans=high
        while(low<=high):
            mid=(low+high)>>1
            if (self.Valid(bloomDay,mid,k)>=m):
                ans=mid
                high=mid-1
            else:
                low=mid+1
            
        return ans
